contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -128,24 +140,347 @@ contract OETHVaultCore is VaultCore {
             "Redeem amount lower than minimum"
         );
 
-        if (IERC20(weth).balanceOf(address(this)) >= amountMinusFee) {
-            // Use Vault funds first if sufficient
-            IERC20(weth).safeTransfer(msg.sender, amountMinusFee);
-        } else {
-            address strategyAddr = assetDefaultStrategies[weth];
-            if (strategyAddr != address(0)) {
-                // Nothing in Vault, but something in Strategy, send from there
-                IStrategy strategy = IStrategy(strategyAddr);
-                strategy.withdraw(msg.sender, weth, amountMinusFee);
-            } else {
-                // Cant find funds anywhere
-                revert("Liquidity error");
-            }
-        }
+        // Is there enough WETH in the Vault available after accounting for the withdrawal queue
+        require(_wethAvailable() >= amountMinusFee, "Liquidity error");
+
+        // Transfer WETH minus the fee to the redeemer
+        IERC20(weth).safeTransfer(msg.sender, amountMinusFee);
 
         // Burn OETH from user (including fees)
         oUSD.burn(msg.sender, _amount);
 
+        // Prevent insolvency
         _postRedeem(_amount);
     }
+
+    /**
+     * @notice Request an asynchronous withdrawal of WETH in exchange for OETH.
+     * The OETH is burned on request and the WETH is transferred to the withdrawer on claim.
+     * This request can be claimed once the withdrawal queue's `claimable` amount
+     * is greater than or equal this request's `queued` amount.
+     * There is no minimum time or block number before a request can be claimed. It just needs
+     * enough WETH liquidity in the Vault to satisfy all the outstanding requests to that point in the queue.
+     * OETH is converted to WETH at 1:1.
+     * @param _amount Amount of OETH to burn.
+     * @param requestId Unique ID for the withdrawal request
+     * @param queued Cumulative total of all WETH queued including already claimed requests.
+     */
+    function requestWithdrawal(uint256 _amount)
+        external
+        whenNotCapitalPaused
+        nonReentrant
+        returns (uint256 requestId, uint256 queued)
+    {
+        // The check that the requester has enough OETH is done in to later burn call
+
+        requestId = withdrawalQueueMetadata.nextWithdrawalIndex;
+        queued = withdrawalQueueMetadata.queued + _amount;
+
+        // Store the next withdrawal request
+        withdrawalQueueMetadata.nextWithdrawalIndex = SafeCast.toUint128(
+            requestId + 1
+        );
+        // Store the updated queued amount which reserves WETH in the withdrawal queue
+        // and reduces the vault's total assets
+        withdrawalQueueMetadata.queued = uint128(queued);
```

Comment:
⚪ We should safe cast the other casts in this file.

Reply:
Agree. Done